### ISSUE TITLE
[New Model]: ByteDance-Seed/BAGEL-7B-MoT

### ISSUE THREAD

**Issue by XueSongTap** [vllm-project/vllm#18793]
### The model to consider.

https://huggingface.co/ByteDance-Seed/BAGEL-7B-MoT

### The closest model vllm already supports.

https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/qwen2_5_vl.py

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by Dexterai** [comment]
hi, I'm interested in contributing but am new to the project, can I work on this ?

**Comment by princepride** [comment]
@DarkLight1337 it seems that the current vLLM framework does not support this kind of multi-modal input and multi-modal output model, right?

**Comment by DarkLight1337** [comment]
Sorry @Dexterai I missed your reply.

We only have minimal support for this kind of model. For now, the pooling task can be used to support multimodal outputs, check out the example of Prithvi-EO model: https://github.com/vllm-project/vllm/blob/main/examples/offline_inference/prithvi_geospatial_mae.py

**Comment by princepride** [comment]
So it only support offline inference?

**Comment by DarkLight1337** [comment]
Yes, only offline inference

**Comment by Dexterai** [comment]
@DarkLight1337  That's ok.  Can vllm now support images or video generation? This model support both image understanding and generating.

**Comment by DarkLight1337** [comment]
> Sorry [@Dexterai](https://github.com/Dexterai) I missed your reply.
> 
> We only have minimal support for this kind of model. For now, the pooling task can be used to support multimodal outputs, check out the example of Prithvi-EO model: https://github.com/vllm-project/vllm/blob/main/examples/offline_inference/prithvi_geospatial_mae.py

See this

**Comment by princepride** [comment]
@Dexterai are you working on this model, I am also interested in this model

**Comment by Dexterai** [comment]
@princepride I give it up, vllm now don't support multimodal outputs. and bagel supports image and video output.

**Comment by princepride** [comment]
@DarkLight1337 As you said, we can only implement the offline inference version of this model, but I am very interested in the implementation of this model. Can you assign this model to me?

**Comment by princepride** [comment]
@Isotr0py Sorry to bother you, if the model's weight were split to two different model and the original project code use hard code to load the model's weight, what should I do, to specific, I want know how vLLM auto load the model's weight file from huggingface repo.
For exmaple (from source code):
```
...
...
from safetensors.torch import load_file as load_sft
def load_ae(local_path: str) -> AutoEncoder:
    ae_params = AutoEncoderParams(
            resolution=256,
            in_channels=3,
            downsample=8,
            ch=128,
            out_ch=3,
            ch_mult=[1, 2, 4, 4],
            num_res_blocks=2,
            z_channels=16,
            scale_factor=0.3611,
            shift_factor=0.1159,
    )

    # Loading the autoencoder
    ae = AutoEncoder(ae_params)

    if local_path is not None:
        sd = load_sft(local_path)
        missing, unexpected = ae.load_state_dict(sd, strict=False, assign=True)
        print_load_warning(missing, unexpected)
    return ae, ae_params

# load the vae model
vae_model, vae_config = load_ae(local_path=os.path.join(args.model_path, "ae.safetensors"))

# load the ema model
model = load_checkpoint_and_dispatch(
        model,
        checkpoint=os.path.join(args.model_path, "ema.safetensors"),
        device_map=device_map,
        offload_buffers=True,
        offload_folder="offload",
        dtype=torch.bfloat16,
        force_hooks=True,
    ).eval()
...
```

**Comment by DarkLight1337** [comment]
I think in that case it may be necessary to create a custom HF Hub repo that uses `model.safetensors.index.json` to map the weights to the file locations. (Unless you can get the official repo authors to update their repo)

**Comment by zhangzef** [comment]
@princepride Hi! Thank you for your contribution. And is there any further progress now, or is there anything I can do to help? I'm also very interested in this job!

**Comment by princepride** [comment]
This week seed already merged my pr to huggingface hub, but as you know, this model is very complex, so I still need sometime

**Comment by princepride** [comment]
@christian-pinto hi, thank you implement the `PrithviGeoSpatialMAE`, I am now working a model also have the image as output so I decided to study your implement, but I found I can not run [vllm](https://github.com/vllm-project/vllm/tree/main)/[examples](https://github.com/vllm-project/vllm/tree/main/examples)/[offline_inference](https://github.com/vllm-project/vllm/tree/main/examples/offline_inference) whatever I do. Can you give me some example code to run the `PrithviGeoSpatialMAE`, thank you so much. 😊

**Comment by christian-pinto** [comment]
> [@christian-pinto](https://github.com/christian-pinto) hi, thank you implement the `PrithviGeoSpatialMAE`, I am now working a model also have the image as output so I decided to study your implement, but I found I can not run [vllm](https://github.com/vllm-project/vllm/tree/main)/[examples](https://github.com/vllm-project/vllm/tree/main/examples)/[offline_inference](https://github.com/vllm-project/vllm/tree/main/examples/offline_inference) whatever I do. Can you give me some example code to run the `PrithviGeoSpatialMAE`, thank you so much. 😊

Hi @princepride please take a look at the code in #20577. You should be able to run the script in the offline_inference examples.

Use [this input file](https://huggingface.co/spaces/ibm-nasa-geospatial/Prithvi-EO-2.0-Sen1Floods11-demo/blob/main/examples/India_900498_S2Hand.tif) and run as in the below snippet:
```bash
python examples/offline_inference/prithvi_geospatial_mae.py \
             --data_file=/path/to/India_900498_S2Hand.tif
```

You should see the files generated in `./output`

**Comment by christian-pinto** [comment]
Also, I am working on a proposal for generating images or other types of data out of the hidden states of the model. I should have a new PR open this week or early next week.


See here: https://github.com/vllm-project/vllm/issues/16052

**Comment by simon-mo** [comment]
@princepride please let us know if you need any help/blocked on anything else.

**Comment by princepride** [comment]
Okay, I'm really sorry it's taken so long to implement this model. I've
been really busy with work lately, and I probably won't have time to work
on it until next month. Previously, when I ran into issues while
implementing this model, I would ask Roger Wang for help.
Lastly, I apologize again for the delay in developing this model.

**Comment by rbao2018** [comment]
mark👀

**Comment by zhangzef** [comment]
I'm not particularly familiar with model inference acceleration. I have a humble opinion: Could we only accelerate the transformer block of the bagel without accelerating the encoding and decoding parts of the multimodal input and output? Wouldn't this be easier to achieve?

**Comment by princepride** [comment]
Of course, in fact, there is no acceleration strategy for the pooling model in vLLM. I was busy with work some time ago, but fortunately, starting from this week (to be more precise, starting from today), I now have time to start re-implementing Bagel. I am very sorry for the delay caused. In fact, the biggest difficulty in implementing Bagel is that it is a model written without any dependence on transformers. This leads to the fact that I need to find a way to figure out how many parameters in this model need to be configured with config during adaptation, because vLLM is highly bound to transformers in the preprocessing and model initialization parts. Of course, I am a little confused when reading the inference code of Bagel's original model, because their image kv cache is actually calculated directly instead of splicing it with the text input in the embedding part and then doing prefill calculation like in transformers and vLLM.

**Comment by zhangzef** [comment]
I understand that your question might be that bagel calculates the kv cache of the image separately? During the reasoning process, bagel does not concatenate with the previous content but only pre-stores the kv cache of the context to accelerate the reasoning process.

**Comment by princepride** [comment]
Yes. In the vLLM framework, the image tokens—processed by the vision module—are directly inserted into placeholders within the text input. The prefill process then calculates the KV cache for both the image and text together. I believe this is a necessary step, because vLLM currently manages the KV cache by calculating a combined hash value from both the image and the input. Therefore, it seems I won't be able to maintain the original code logic during the Bagel inference process.

**Comment by zhangzef** [comment]
OK! I understand. Actually, bagel can also be calculated using the prefill method. You can look at its forward function. It can also accept a set of encoded vae tokens, vit tokens, and text tokens. It will preprocess these tokens into a format that can be input into the transformer during the forward process and then uniformly send them to the language model for forward. However, in the inferencer, the author used a pre-stored kv cache to accelerate inference.

However, its data preprocessing format is rather complex. You may need to carefully observe the data preprocessing methods in its dataset_base.py and interleave_datasets folders to make your data format match its input.

**Comment by zhangzef** [comment]
And it uses a `packed_sequence` variable to store the token embedding of the prefill.

**Comment by tingxueronghua** [comment]
awesome work and I think researchers will all love it! Is there any progress recently?

**Comment by princepride** [comment]
Sorry for the delay. I encountered some difficulties in developing bagel before, so I turned to the development of several other models. Yesterday, these models were all merged, and I turned back to the development of bagel.
